File: src/main/java/org/lanternpowered/server/advancement/criteria/trigger/LanternFilteredTriggerBuilder.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.advancement.criteria.trigger;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import org.spongepowered.api.advancement.criteria.trigger.FilteredTrigger;
import org.spongepowered.api.advancement.criteria.trigger.FilteredTriggerConfiguration;
import org.spongepowered.api.advancement.criteria.trigger.Trigger;

@SuppressWarnings({"unchecked", "ConstantConditions", "NullableProblems"})
public class LanternFilteredTriggerBuilder<C extends FilteredTriggerConfiguration> implements FilteredTrigger.Builder<C> {

    private C config;
    private Trigger<C> type;

    @Override
    public <T extends FilteredTriggerConfiguration> FilteredTrigger.Builder<T> type(Trigger<T> type) {
        checkNotNull(type, "type");
        this.type = (Trigger<C>) type;
        return (FilteredTrigger.Builder<T>) this;
    }

    @Override
    public FilteredTrigger.Builder<C> config(C config) {
        checkNotNull(config, "config");
        this.config = config;
        return this;
    }

    @Override
    public FilteredTrigger<C> build() {
        checkState(this.type != null, "The type must be set");
        checkState(this.config != null, "The config must be set");
        return new LanternFilteredTrigger<>(this.type, this.config);
    }

    @Override
    public FilteredTrigger.Builder<C> from(FilteredTrigger<C> value) {
        this.config = value.getConfiguration();
        this.type = value.getType();
        return this;
    }

    @Override
    public FilteredTrigger.Builder<C> reset() {
        this.config = null;
        this.type = null;
        return this;
    }
}

File: src/main/kotlin/org/lanternpowered/server/advancement/criteria/trigger/LanternFilteredTriggerBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.advancement.criteria.trigger

import org.spongepowered.api.advancement.criteria.trigger.FilteredTrigger
import org.spongepowered.api.advancement.criteria.trigger.FilteredTriggerConfiguration
import org.spongepowered.api.advancement.criteria.trigger.Trigger

@Suppress("UNCHECKED_CAST")
class LanternFilteredTriggerBuilder<C : FilteredTriggerConfiguration> : FilteredTrigger.Builder<C> {

    private var config: C? = null
    private var type: Trigger<C>? = null

    override fun <T : FilteredTriggerConfiguration> type(type: Trigger<T>): FilteredTrigger.Builder<T> {
        this.type = type as Trigger<C>
        return this as LanternFilteredTriggerBuilder<T>
    }

    override fun config(config: C): FilteredTrigger.Builder<C> = apply { this.config = config }

    override fun build(): FilteredTrigger<C> {
        val type = checkNotNull(this.type) { "The type must be set" }
        val config = checkNotNull(this.config) { "The config must be set" }
        return LanternFilteredTrigger(type, config)
    }

    override fun from(value: FilteredTrigger<C>): FilteredTrigger.Builder<C> = apply {
        this.config = value.configuration
        this.type = value.type
    }

    override fun reset(): FilteredTrigger.Builder<C> = apply {
        this.config = null
        this.type = null
    }
}
